### PR TITLE
http3: applied cluster buffer limits to http/3

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: Low (http/3 only)
Testing: existing flow control tests pass
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/quic/client_connection_factory_impl.cc [hunk 1]
```diff
@@ -62,12 +64,10 @@ createQuicNetworkConnection(Http::PersistentQuicInfo& info, Event::Dispatcher& d
 
   ASSERT(!info_impl->supported_versions_.empty());
   // QUICHE client session always use the 1st version to start handshake.
-  // TODO(alyssawilk) pass in ClusterInfo::perConnectionBufferLimitBytes() for
-  // send_buffer_limit instead of using 0.
   auto ret = std::make_unique<EnvoyQuicClientSession>(
       info_impl->quic_config_, info_impl->supported_versions_, std::move(connection),
       info_impl->server_id_, info_impl->crypto_config_.get(), &static_info.push_promise_index_,
-      dispatcher, /*send_buffer_limit=*/0);
+      dispatcher, info_impl->buffer_limit_);
```
Comment (danzh2010) [inline source/common/quic/client_connection_factory_impl.cc:70]: Mind adding some conn_pool tests for this?

**Review by danzh2010 — COMMENTED** [review]
Looks good overall along aside a request for testing.

File: source/common/quic/client_connection_factory_impl.cc [hunk 2]
```diff
@@ -19,24 +19,26 @@ getContext(Network::TransportSocketFactory& transport_socket_factory) {
...
           std::make_unique<EnvoyQuicProofVerifier>(getContext(transport_socket_factory)),
-          std::make_unique<EnvoyQuicSessionCache>(time_source))) {
+          std::make_unique<EnvoyQuicSessionCache>(time_source))),
+      buffer_limit_(buffer_limit) {
   quiche::FlagRegistry::getInstance();
 }
 
 namespace {
 // TODO(alyssawilk, danzh2010): This is mutable static info that is required for the QUICHE code.
+
```
Comment (danzh2010) [inline source/common/quic/client_connection_factory_impl.cc:41]: irrelevant?

File: test/common/http/http3/conn_pool_test.cc [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  NiceMock<Server::Configuration::MockTransportSocketFactoryContext> context_;
+  Quic::QuicClientTransportSocketFactory factory_{
+      std::unique_ptr<Envoy::Ssl::ClientContextConfig>(new NiceMock<Ssl::MockClientContextConfig>),
+      context_};
+  ConnectionPool::InstancePtr pool_;
+};
+
+TEST_F(Http3ConnPoolImplTest, CreationWithBufferLimits) {
+  EXPECT_CALL(mock_host->cluster_, perConnectionBufferLimitBytes);
+  initialize();
```
Comment (danzh2010) [inline test/common/http/http3/conn_pool_test.cc:46]: Thanks for adding the test! Is it possible to test createQuicNetworkConnection() actually pass the buffer limit to EnvoyQuicClientSession?

**Comment by alyssawilk** [comment]
@snowp mind doing !googler pass?

File: test/common/http/http3/conn_pool_test.cc [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    new Event::MockSchedulableCallback(&dispatcher_);
+    Network::ConnectionSocket::OptionsSharedPtr options;
+    Network::TransportSocketOptionsSharedPtr transport_options;
+    pool_ = allocateConnPool(dispatcher_, random_, host_, Upstream::ResourcePriority::Default,
+                             options, transport_options, state_, simTime());
+  }
+
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  std::shared_ptr<Upstream::MockClusterInfo> cluster_{new NiceMock<Upstream::MockClusterInfo>()};
+  Upstream::MockHost* mock_host = new NiceMock<Upstream::MockHost>;
```
Comment (snowp) [inline test/common/http/http3/conn_pool_test.cc:31]: Naming convention seems off for this field. Maybe just combine this and host_ into one field and avoid the raw pointer?

**Review by snowp — CHANGES_REQUESTED** [review]
Just one comment, otherwise this LGTM

File: test/common/http/http3/conn_pool_test.cc (hunk 3 above)
Comment (danzh2010) [reply test/common/http/http3/conn_pool_test.cc:47]: initialize() seems only creates the pool object instead of the client.
I'm concerned about future regression in pool_->instantiateActiveClient() that the limit didn't get picked up.

**Comment by alyssawilk** [comment]
Done.  Any other requests?

File: test/common/quic/client_connection_factory_impl_test.cc [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  Quic::QuicClientTransportSocketFactory factory_{
+      std::unique_ptr<Envoy::Ssl::ClientContextConfig>(new NiceMock<Ssl::MockClientContextConfig>),
+      context_};
+
+  uint32_t highWatermark(EnvoyQuicClientSession* session) {
+    return session->write_buffer_watermark_simulation_.highWatermark();
+  }
+};
+
+TEST_F(QuicNetworkConnectionTest, BufferLimits) {
```
Comment (danzh2010) [inline test/common/quic/client_connection_factory_impl_test.cc:35]: Nice!

**Review by danzh2010 — APPROVED** [review]
LGTM

**Comment by snowp** [comment]
Needs main merge + CI fix, code LG

File: source/common/quic/envoy_quic_proof_verifier.h [hunk 6]
```diff
@@ -12,7 +12,9 @@ namespace Quic {
 class EnvoyQuicProofVerifier : public EnvoyQuicProofVerifierBase {
 public:
   EnvoyQuicProofVerifier(Envoy::Ssl::ClientContextSharedPtr&& context)
-      : context_(std::move(context)) {}
+      : context_(std::move(context)) {
+    ASSERT(context_.get());
+  }
```
Comment (snowp) [inline source/common/quic/envoy_quic_proof_verifier.h:18]: Does this change belong here? Not opposed to it, just seems unrelated to the main change in this PR

Comment (alyssawilk) [reply source/common/quic/envoy_quic_proof_verifier.h:18]: it was a follow-up to a change requested in another PR.
I can split it into its own thing, but didn't think it was worth the CI cost :-P

Comment (snowp) [reply source/common/quic/envoy_quic_proof_verifier.h:18]: Fine with me as long as we're aware that we're doing it!

**Review by snowp — DISMISSED** [review]

**Review by snowp — APPROVED** [review]
